napari/_qt/qt_viewer.py

```diff
@@ -226,8 +227,26 @@ def screenshot(self):
             upper-left corner of the rendered region.
         """
         img = self.canvas.native.grabFramebuffer()
+        if path is not None:
+            imsave(path, QImg2array(img))  # scikit-image imsave method
         return QImg2array(img)
 
+    def _save_screenshot(self):
+        """Save screenshot of currently displayed screen to file.
+
+        Returns
+        -------
+        filename : str
+            File path of the saved screenshot.
+        """
+        filename, _ = QFileDialog.getSaveFileName(
```

Comment:
Is there a way to autopopulate the save name space with something that has a `png` extension, or a clean way for us to automatically add a `png` extension if someone just types in a string.

I just tried to take a screenshot, typed in something like `astro` and then got this error

```python-traceback
(base) czirwc1macos2701:napari nsofroniew$ python examples/add_image.py
Traceback (most recent call last):
  File "/Users/nsofroniew/GitHub/napari/napari/_qt/qt_viewer.py", line 248, in _save_screenshot
    self.screenshot(path=filename)
  File "/Users/nsofroniew/GitHub/napari/napari/_qt/qt_viewer.py", line 231, in screenshot
    imsave(path, QImg2array(img))  # scikit-image imsave method
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/site-packages/skimage/io/_io.py", line 131, in imsave
    return call_plugin('imsave', fname, arr, plugin=plugin, **plugin_args)
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/site-packages/skimage/io/manage_plugins.py", line 210, in call_plugin
    return func(*args, **kwargs)
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/site-packages/imageio/core/functions.py", line 302, in imwrite
    writer = get_writer(uri, format, "i", **kwargs)
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/site-packages/imageio/core/functions.py", line 226, in get_writer
    "Could not find a format to write the specified file " "in mode %r" % mode
ValueError: Could not find a format to write the specified file in mode 'i'
```

If I typed `astro.png` it worked fine. Maybe @tlambert03 has a suggestion on how to do this well

Reply:
If we do pathname manipulations, I'd like that to be something that happens as a convenience *only* to strings returned through GUI interaction. 

There are two main options:
1. Append ".png" to the end of whatever string is provided. Possible disadvantage: this does mean it is possible to wind up with filenames like "file.png.png" or "file.tif.png".

2. Use `os.path.splitext()` on whatever string is provided and replace the extension with ".png". Possible disadvantage: this means it's possible to lose some filename information you might have intended to keep, for instance if you type "file.name" expecting to get "file.name.png" instead you'll wind up with "file.png". 

**I'm leaning towards option 1, appending ".png" to whatever string was returned from the file menu "Screenshot" option.** 

I prefer this because:
* there's no possible loss of information that was typed by the user
* there is no pathname manipulation if you are using the API for scripting (I would find that super unintuitive)

Reply:
I definitely agree with "convenience only to strings returned through GUI interaction". I'd prefer to use the built-in Qt Dialog methods for this though, sorry I didn't link to them earlier (I had to do a bit of reading up on them now before replying as I wasn't too familiar with them before). They have a concept called `NameFilters` - see here https://doc.qt.io/qt-5/qfiledialog.html#setNameFilters, that allows the desired behaviour. I think adding `filter="Images (*.png *tif *.tiff)",` to `QFileDialog.getSaveFileName` will achieve what we want. If no extension is provided then `.png` will be added. If any one of the extensions listed are provided they will be used, and if a different extension is provided you will get a standard warning and an option to save with both `.xyz.png` or just `.xyz` if that is what you had typed. You can then drop the `+.png`  you were adding.

Can you also adjust the tooltip to say something like `as png by default` or drop the mention of `png` entirely

Reply:
I can't make this work, sorry. I've tried several different permutations on this idea but it seems that the filters only affect what is visible to you in the dialog box, not the value of what is returned. This seems relevant - https://bugreports.qt.io/browse/QTBUG-27186

There is also supposed to be a `setDefaultSuffix`, but I can't get that working properly with getSaveFilename either :(

Reply:
No worries, there are sometimes bugs / weird things with Qt, and sometimes they vary depending on what platform you're on!!! I can take a look at this tomorrow, or maybe @tlambert03 has some experience here and can also take a look (If you can @tlambert03 it would be appreciated as I am quite busy tomorrow). If either of us can get it working we'll push the changes directly to this PR if that's ok, if we can't then we can go with your solution or one where we implement a little bit of this logic ourselves. Thanks for trying!!

Reply:
@GenevieveBuckley I've gotten this working for me now, I just pushed the changes directly to the branch. Can you check them out? There was a typo in what I pasted above `(*.png *tif *.tiff)` vs `(*.png *.tif *.tiff)` which will have caused all sorts of problems if not caught, my apologies. I've included a gif in the main thread of the behavior for me.

Reply:
No, this fails for me. 

The problem must be platform dependent (running Linux Ubuntu here), it's the same as when I tried earlier. Essentially, if you give it a filename without an extension it won't automatically add one for you like it is supposed to do. The Qt dialog is definitely *supposed* to work the way you expect, but it seems too buggy to rely on here.

Here's my system and napari info:
```
napari: 0.2.11+17.g66fcdc79
Platform: Linux-5.3.0-28-generic-x86_64-with-debian-buster-sid
Python: 3.7.4 (default, Aug 13 2019, 20:35:49) [GCC 7.3.0]
Qt: 5.14.0
PySide2: 5.14.0
VisPy: 0.6.4
NumPy: 1.18.1
SciPy: 1.4.0.dev0+1566e8e
scikit-image: 0.16.2
Dask: 2.10.0

GL version: 4.6.0 NVIDIA 435.21
MAX_TEXTURE_SIZE: 32768
```

`pytest napari/_qt/_tests/test_qt_viewer.py::test_save_screenshot` fails on my system.
The relevant traceback is here (imageio fails because it wasn't given a filename with a file extension):

```python-traceback
Traceback (most recent call last):
  File "/home/genevieve/Documents/GitHub/napari/napari/_qt/qt_viewer.py", line 249, in _save_screenshot
    self.screenshot(path=filename)
  File "/home/genevieve/Documents/GitHub/napari/napari/_qt/qt_viewer.py", line 236, in screenshot
    imsave(path, QImg2array(img))  # scikit-image imsave method
  File "/home/genevieve/anaconda3/envs/napari-dev/lib/python3.7/site-packages/skimage/io/_io.py", line 131, in imsave
    return call_plugin('imsave', fname, arr, plugin=plugin, **plugin_args)
  File "/home/genevieve/anaconda3/envs/napari-dev/lib/python3.7/site-packages/skimage/io/manage_plugins.py", line 210, in call_plugin
    return func(*args, **kwargs)
  File "/home/genevieve/anaconda3/envs/napari-dev/lib/python3.7/site-packages/imageio/core/functions.py", line 302, in imwrite
    writer = get_writer(uri, format, "i", **kwargs)
  File "/home/genevieve/anaconda3/envs/napari-dev/lib/python3.7/site-packages/imageio/core/functions.py", line 226, in get_writer
    "Could not find a format to write the specified file " "in mode %r" % mode
ValueError: Could not find a format to write the specified file in mode 'i'
```

Reply:
ah, that's really no good - and that issue you link to above really does seem unresolved. Agreed we shouldn't have something that breaks on linux. Out of curiosity what is the behavior like if you put in a "bad" file extension, say `jpg` in the context of the current PR, do you get a warning and a request to add `png` or does it do nothing?

I'm inclined to keep the current behavior and then add in your previous catch in case it fails, that should make the best of this not perfect situation. i can submit a fix soon

Reply:
With jpeg I get an error say pillow won't support an alpha channel (I just opened a png with the camera data in it, didn't think that should have had an alpha channel to begin with).

Reply:
Ok, what that means is that the dialog option is completely ignored for you on linux, for me on mac it catches it earlier. The screenshot does have an alpha value as it allows us to capture opacity from vispy in what I think is the best way possible, but I might be wrong - if the alpha doesn't get used we could drop.

Reply:
> Ok, what that means is that the dialog option is completely ignored for you on linux, for me on mac it catches it earlier. 

No, I don't think this is strictly accurate. The "filters" dialog option does get used on Linux - in the dialog box that pops up I see only folders and files that match the filter (so no text files, etc. are visible to me there). What it doesn't do is automatically use the first file extension in that filters list if the user types in a filename without any file extension.

Reply:
Thanks for the clarification!